docs/development/go/file-io/index.md

```diff
@@ -0,0 +1,720 @@
+    }
+
+    destination, err := os.Create(dst)
+    if err != nil {
+        return err
+    }
+    defer destination.Close()
+
+    buf := make([]byte, BUFFERSIZE)
+    for {
+        n, err := source.Read(buf)
+        if err != nil && err != io.EOF {
+            return err
+        }
+        if n == 0 {
+            break
+        }
+
+        if _, err := destination.Write(buf[:n]); err != nil {
+            return err
+        }
+    }
+    return err
+}
+
+func main() {
+    if len(os.Args) != 4 {
+        fmt.Printf("usage: %s source destination BUFFERSIZE\n", filepath.Base(os.Args[0]))
+        return
+    }
+
+    source := os.Args[1]
+    destination := os.Args[2]
+    BUFFERSIZE, _ = strconv.ParseInt(os.Args[3], 10, 64)
+
+    fmt.Printf("Copying %s to %s\n", source, destination)
+    err := Copy(source, destination, BUFFERSIZE)
+    if err != nil {
+        fmt.Printf("File copying failed: %q\n", err)
+    }
+}
+{{< /file >}}
+
+The good thing with `fileCopy.go` is that it allows you to set the size of the buffer that will be used during the copy process – as it happens in Go, that buffer is implemented using a *byte slice* named `buf`. The copy takes place in the `Copy()` function that keeps reading the input file using the required amount of `Read()` calls and writes it using the required amount of `Write()` calls. The `Copy()` function performs lots of tests to make sure that the source file exists and is a regular file and that the destination file does not exist.
+
+The output of `fileCopy.go` will resemble the following:
+
+    go run fileCopy.go /tmp/data.txt /tmp/newText 16
+{{< output >}}
+Copying /tmp/data.txt to /tmp/newText
+{{< /output >}}
+
+### Implementing cat in Go
+
+In this section we will implement the core functionality of the `cat(1)` command line utility in Go.
+
+{{< file "./cat.go" go >}}
+package main
+
+import (
+    "bufio"
+    "fmt"
+    "io"
+    "os"
+)
+
+func printFile(filename string) error {
+    f, err := os.Open(filename)
+    if err != nil {
+        return err
+    }
+    defer f.Close()
+    scanner := bufio.NewScanner(f)
+    for scanner.Scan() {
+        io.WriteString(os.Stdout, scanner.Text())
+        io.WriteString(os.Stdout, "\n")
+    }
+    return nil
+}
+
+func main() {
+    filename := ""
+    arguments := os.Args
+    if len(arguments) == 1 {
+        io.Copy(os.Stdout, os.Stdin)
+        return
+    }
+
+    for i := 1; i < len(arguments); i++ {
+        filename = arguments[i]
+        err := printFile(filename)
+        if err != nil {
+            fmt.Println(err)
+        }
+    }
+}
+{{< /file >}}
+
+If you execute `cat.go` without any command line arguments, then the utility will just copy from standard input to standard output using the `io.Copy(os.Stdout, os.Stdin)` statement. However, if there are command-line arguments, then the program will process them all in the same order that they were given using the `printFile()` function.
```

Comment:
What are the command line arguments that are possible based on this code sample? If we can't use any command line arguments using this sample alone, can you go into a little more detail to explain that and talk about how it could be achieved?

Reply:
`cat.go` works just like the `cat(1)` utility. If you use command line arguments, which should be valid file paths, it will read those files and print their contents. If you use no command line arguments, it will print standard input (user input) to standard output until you press Control+D.
In the given example, `cat.go` processes a file.
@rsyracuse Please let me know if the use of `cat.go` is clear now.

Reply:
That clears it up. Thanks!